### PR TITLE
Fix Cody Chat UI rendering reconciliation

### PR DESCRIPTION
**Pull request by vovakulikov**
Fixes part of https://linear.app/sourcegraph/issue/CODY-3943/optimize-performance-for-large-chat-windows-in-cody

It looks like since https://github.com/sourcegraph/cody/pull/5036 we've lost memoization in a few places. As a result, our chat UI doesn't perform well with more than 4-5 messages (depending on how long these chat messages are)

This PR tries to improve it by fixing memoization on expensive components (Transcript, HumanMessage, ...etc)

| Before | After |
| ------- | ------- |
| <video src="https://github.com/user-attachments/assets/79e7b93c-ac82-4a8c-9b2e-6c712af3fb2c" /> | <video src="https://github.com/user-attachments/assets/223b3513-0f30-47cc-b24c-c54f4861329c" /> | 

This PR also fixes a problem with one-box intention detection; previously, it produced a state where we didn't have a follow-up message UI during intent resolution. 

## Test plan
- Check with the React profile that we don't render components during message response in places where we don't need to re-render anything
- Check that intent detection in one-box doesn't produce UI flashes

### CODE REVIEW

**Review by abeatrix — APPROVED** [review]
Tested locally and confirmed the states were not re-render.

I was looking into a similar issue for the editor state. Would appreciate your review on it as well since I think there are a few things I can clean up on my PR after yours is merged 😄  https://github.com/sourcegraph/cody/pull/5847

**Comment by abeatrix** [comment]
(Not a blocker) would be great to have a test covering transcript rendering for regression if possible 😄

**Review by thenamankumar — APPROVED** [review]

File: vscode/webviews/utils/debugger.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { useEffect, useRef } from 'react'
+
+export function useTracePropsUpdate(props: any) {
+    const prev = useRef(props)
+
+    useEffect(() => {
+        const changedProps = Object.entries(props).reduce(
+            (ps, [k, v]) => {
```
Comment (dominiccooney) [inline vscode/webviews/utils/debugger.ts:8]: Our lints for PromptString rely on only using `ps` to refer to the PromptString formatter. This ensures we never alias `ps`, which can be used to defeat the static checks for correct PromptString usage.

We could write a more complicated check that understands JavaScript scoping to allow the use you have here, or only lint files which import "the" `ps`, but until we do that, can you rename these variables to anything else?

**Review by dominiccooney — CHANGES_REQUESTED** [review]
We need a trivial renaming to keep the PromptString lint results accurate.

File: vscode/webviews/utils/debugger.ts (hunk 1 above)
Comment (vovakulikov) [reply vscode/webviews/utils/debugger.ts]: Sure, renamed it in my last commit

File: vscode/webviews/utils/debugger.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { useEffect, useRef } from 'react'
+
+export function useTracePropsUpdate(props: any) {
```
Comment (dominiccooney) [inline vscode/webviews/utils/debugger.ts:3]: Let's add a comment here that this is for debugging with advice about how to use it.

From time to team people do mass-deletions based on use. You probably intend for this to stick around.

File: vscode/webviews/chat/Transcript.tsx [hunk 3]
```diff
@@ -389,6 +389,17 @@ const TranscriptInteraction: FC<TranscriptInteractionProps> = memo(props => {
     )
 }, isEqual)
 
+function useMutatedValue<T>(value?: T): [MutableRefObject<T | undefined>, setValue: (value: T) => void] {
```
Comment (dominiccooney) [inline vscode/webviews/chat/Transcript.tsx:392]: You should document that this should not be used for rendering and won't cause re-rendering.

File: vscode/webviews/chat/Transcript.tsx [hunk 4]
```diff
@@ -180,7 +180,7 @@ const TranscriptInteraction: FC<TranscriptInteractionProps> = memo(props => {
         smartApplyEnabled,
     } = props
 
-    const [intentResults, setIntentResults] = useState<
+    const [intentResults, setIntentResults] = useMutatedValue<
```
Comment (dominiccooney) [inline vscode/webviews/chat/Transcript.tsx:183]: Why isn't simply `useRef` enough here?

**Review by dominiccooney — APPROVED** [review]
Approving for the `ps` fixes, thank you. I have not reviewed the rest in detail but I have some questions and suggestions inline.

I think the message output stuttering has a different cause, see https://github.com/sourcegraph/cody/issues/5296#issuecomment-2402410782

**Pull request merged**